File: apps/mobile/smarttaxi_app/lib/features/shared/stand_outcome.dart
import '../../l10n/app_localizations.dart';

class StandOutcome {
  const StandOutcome({required this.id, required this.status, this.reason});
  final String id;
  final String status;
  final String? reason;

  factory StandOutcome.fromJson(Map<String, dynamic> json) {
    if (json['id'] is! String ||
        json['status'] is! String ||
        (json['reason'] != null && json['reason'] is! String)) {
      throw const FormatException('Invalid stand outcome');
    }
    return StandOutcome(
        id: json['id'] as String,
        status: json['status'] as String,
        reason: json['reason'] as String?);
  }
}

enum StandOutcomeNotice {
  closedDriver,
  closedRider,
  expired,
  declined,
  cancelled,
  boarded,
  carLeft,
  leftArea,
  noSignal,
  handedOver,
  acceptedOrder,
  driverLeft,
  riderUnavailable;

  String text(AppLocalizations l10n) => switch (this) {
        closedDriver => l10n.standOutcomeClosedDriver,
        closedRider => l10n.standOutcomeClosedRider,
        expired => l10n.standOutcomeExpired,
        declined => l10n.standOutcomeDeclined,
        cancelled => l10n.standOutcomeCancelled,
        boarded => l10n.standOutcomeBoarded,
        carLeft => l10n.standOutcomeCarLeft,
        leftArea => l10n.standOutcomeLeftArea,
        noSignal => l10n.standOutcomeNoSignal,
        handedOver => l10n.standOutcomeHandedOver,
        acceptedOrder => l10n.standOutcomeAcceptedOrder,
        driverLeft => l10n.standOutcomeDriverLeft,
        riderUnavailable => l10n.standOutcomeRiderUnavailable,
      };
}

StandOutcomeNotice? standOutcomeNotice(StandOutcome? outcome, String id,
    {bool driver = false}) {
  final fallback = driver
      ? StandOutcomeNotice.driverLeft
      : StandOutcomeNotice.riderUnavailable;
  if (outcome == null || outcome.id != id) return fallback;
  if (['WAITING', 'BOARDING', 'PENDING', 'CONFIRMED']
      .contains(outcome.status)) {
    return null;
  }
  if (driver) {
    return switch (outcome.reason) {
      'STAND_CLOSED' => StandOutcomeNotice.closedDriver,
      'LEFT_AREA' => StandOutcomeNotice.leftArea,
      'NO_SIGNAL' => StandOutcomeNotice.noSignal,
      'GAVE_TURN' => StandOutcomeNotice.handedOver,
      'ACCEPTED_ORDER' => StandOutcomeNotice.acceptedOrder,
      _ => outcome.status == 'DEPARTED' ? StandOutcomeNotice.boarded : fallback,
    };
  }
  if (outcome.status == 'EXPIRED') return StandOutcomeNotice.expired;
  if (outcome.status == 'DECLINED') return StandOutcomeNotice.declined;
  if (outcome.status == 'BOARDED') return StandOutcomeNotice.boarded;
  if (outcome.status != 'CANCELLED') return fallback;
  if (outcome.reason == 'STAND_CLOSED') return StandOutcomeNotice.closedRider;
  return outcome.reason != null
      ? StandOutcomeNotice.carLeft
      : StandOutcomeNotice.cancelled;
}

File: apps/web/src/features/shared/standOutcome.mjs
export const standOutcomeMessages = {
  closedDriver: 'Стоянку закрыли. Ваше место в очереди снято. Выберите другую стоянку.',
  closedRider: 'Стоянку закрыли. Бронь снята. Выберите другую стоянку или закажите поездку.',
  expired: 'Время ожидания подтверждения истекло. Бронь снята — можно выбрать другую машину.',
  declined: 'Водитель не подтвердил бронь. Выберите другую машину.',
  cancelled: 'Бронь отменена. Место больше не занято.',
  boarded: 'Посадка завершена. Машина отправилась.',
  carLeft: 'Машина больше не набирает пассажиров. Бронь снята — выберите другую машину.',
  leftArea: 'Вы покинули зону стоянки. Место в очереди снято. Чтобы вернуться, подъедьте к стоянке.',
  noSignal: 'Давно не было координат от вашего устройства. Место снято. Проверьте GPS и встаньте в очередь заново.',
  handedOver: 'Ваше место передано другому водителю.',
  acceptedOrder: 'Вы приняли заказ. Место на стоянке освобождено.',
  driverLeft: 'Ваше место в очереди освобождено.',
  riderUnavailable: 'Эта бронь больше не активна. Обновите данные или выберите другую машину.',
};

export function standOutcomeNotice(outcome, id, driver = false) {
  const fallback = driver ? 'driverLeft' : 'riderUnavailable';
  if (!outcome || outcome.id !== id || typeof outcome.status !== 'string') return fallback;
  if (['WAITING', 'BOARDING', 'PENDING', 'CONFIRMED'].includes(outcome.status)) return null;
  if (driver) {
    if (outcome.reason === 'STAND_CLOSED') return 'closedDriver';
    if (outcome.reason === 'LEFT_AREA') return 'leftArea';
    if (outcome.reason === 'NO_SIGNAL') return 'noSignal';
    if (outcome.reason === 'GAVE_TURN') return 'handedOver';
    if (outcome.reason === 'ACCEPTED_ORDER') return 'acceptedOrder';
    return outcome.status === 'DEPARTED' ? 'boarded' : fallback;
  }
  if (outcome.status === 'EXPIRED') return 'expired';
  if (outcome.status === 'DECLINED') return 'declined';
  if (outcome.status === 'BOARDED') return 'boarded';
  if (outcome.status !== 'CANCELLED') return fallback;
  if (outcome.reason === 'STAND_CLOSED') return 'closedRider';
  return outcome.reason ? 'carLeft' : 'cancelled';
}
